app/scripts/views/settings.js

```diff
@@ -3,114 +3,296 @@
+  'views/settings/display_name',
   'views/mixins/settings-mixin',
+  'views/decorators/allow_only_one_submit',
   'stache!templates/settings'
 ],
-function (Cocktail, Session, FormView, BaseView, AvatarMixin,
-  SettingsMixin, Template) {
+function ($, modal, Cocktail, p, Session, BaseView, AvatarMixin,
+  AvatarView, AvatarChangeView, AvatarCropView, AvatarCameraView, GravatarView,
+  GravatarPermissionsView, CommunicationPreferencesView, ChangePasswordView,
+  DeleteAccountView, DisplayNameView, SettingsMixin, allowOnlyOneSubmit,
+  Template) {
   'use strict';
 
   var t = BaseView.t;
 
-  var View = FormView.extend({
+  var PANEL_VIEWS = [
+    AvatarView,
+    DisplayNameView,
+    CommunicationPreferencesView,
+    ChangePasswordView,
+    DeleteAccountView,
+    AvatarChangeView,
+    AvatarCropView,
+    AvatarCameraView,
+    GravatarView,
+    GravatarPermissionsView
+  ];
+
+  function renderView(view) {
+    return view.render()
+      .then(function (shown) {
+        if (! shown) {
+          view.destroy(true);
+          return;
+        }
+        view.afterVisible();
+
+        return view;
+      });
+  }
+
+  var View = BaseView.extend({
     template: Template,
     className: 'settings',
 
+    FADE_OUT_SETTINGS_MS: 250,
+
     initialize: function (options) {
       options = options || {};
 
       this._able = options.able;
+      this._subViewToShow = options.subView;
+      this._panelViews = options.panelViews || PANEL_VIEWS;
+
+      this.router.on(this.router.NAVIGATE_FROM_SUBVIEW, this._onNavigateFromSubview.bind(this));
     },
 
     context: function () {
       var account = this.getSignedInAccount();
-      var email = account.get('email');
 
       return {
-        email: email,
+        username: account.get('email'),
         showSignOut: ! account.isFromSync(),
         communicationPrefsVisible: this._areCommunicationPrefsVisible()
       };
     },
 
     events: {
-      // validateAndSubmit is used to prevent multiple concurrent submissions.
-      'click #signout': BaseView.preventDefaultThen('validateAndSubmit')
+      'click #signout': BaseView.preventDefaultThen('signOut')
     },
 
-    beforeRender: function () {
-      if (this.relier.get('setting') === 'avatar') {
-        this.navigate('/settings/avatar/change');
-        return false;
+    onProfileUpdate: function () {
+      this._showAvatar();
+    },
+
+    // When we navigate to settings from a subview
+    // close the modal, show any ephemeral messages passed to `navigate`
+    _onNavigateFromSubview: function () {
+      if ($.modal.isActive()) {
+        $.modal.close();
       }
+      this.showEphemeralMessages();
+      this.logScreen();
     },
 
-    afterRender: function () {
-      this.logScreenEvent('communication-prefs-link.visible.' +
-          String(this._areCommunicationPrefsVisible()));
+    showSubView: function (SubView, options) {
+      var self = this;
+      if (self._panelViews.indexOf(SubView) === -1) {
```

Comment:
What is this guarding against? My guess is to protect against showing a view that's not a registered subview. Should a message be logged to the console? Can you add a comment?